### ISSUE TITLE
storage semantics make it hard to integrate with other services

### ISSUE THREAD

**Issue by idrarig** [upspin/upspin#571]
In theory, the Upspin dir and store server interfaces make it possible to build servers that integrate the Upspin filesystem with many other services. (An example made at the end of [this presentation](https://www.youtube.com/watch?v=ENLWEfi0Tkg) is that of Twitter feeds. Not clear whether that was a read-only example, or whether a user could tweet by echoing a string to a text file.)

It seems in practice hard to do, mainly because of the fact that each user
stores blobs of data in their own store server.

Consider the setting:

- alice@example.org main user
- alice+service@example.org suffixed user for the sole purpose of attaching another $service to the Upspin filesystem
- a link from alice@example.org/service to alice+service@example.org/ for Alice's use of service

This seems a plausible set up unless Alice is willing to run more than one instance of upspinfs. 

If Alice wants to write something to $service (which could be, predictably, a file upload, but could be something else like editing the list of tasks in a project in a task management app), the $service's dir server needs to do the following

- fetch blobs from Alice's store server (retrying until the cacheserver has actually uploaded them, which could be a long time if there's a long queue, say, you've just moved an hour-long video to Upspin)
- decrypt those blobs (most likely)
- use those blobs to do the wanted write operations on $service (note these are orphaned in Alice's store server and could be garbage-collected)

This is involved and appears to be the pattern all custom servers (that need to write) need to follow. I am doing this as an experiment into [integrating Google Drive into an Upspin tree](https://github.com/idrarig/upspin).

(As a user, what I would like to do is to specify in my config to "mount" a certain dirserver on /u/alice@example.org/service doing away with suffixed users altogether, and sending blobs for such dir in the $service's store server, not Alice's.)

**Comment by robpike** [comment]
As expressed here, I see the problems you are expressing, at least somewhat. But in the original thinking at least, Twitter (==$service) would be the one providing the service, not Alice. That is, Alice would not set up a Twitter setup herself; the unreasonable idea was that Twitter would provide the Upspin APIs on its server. Alice would just use it, not run it.

Unless I do not understand your criticism, which is likely.

**Comment by idrarig** [comment]
I see; I feel like I'm missing something obvious: doesn't that just move the same complications from Alice to Twitter? Unless data is exposed read-only, that is, which is easy.

I might have partially misunderstood the project: perhaps it was never meant to integrate in read-write mode with other services?

Anyway, I just meant to share the problems I encountered, and what seems to be a workaround. Is it right? Is it wrong? Discouraged? Is it worth it? Is what I tried to pursue a non-goal for the project?

I hope this follow up clarifies things a little. Otherwise feel free to close, I wanted to help, not hinder :)

/cc @adg

**Comment by robpike** [comment]
Well, Twitter is read-write already, and encryption isn't relevant to their service the way it is to the Upspin core. It should in principle be simple to provide a plaintext read-write Upspin service that accesses Twitter data. Of course, they have no interest in that.

To be clear: the idea you're referencing is from the manifesto, not the actual system docs.

**Comment by adg** [comment]
Assuming Twitter was interested in providing a read/write Upspin interface to their system, I think the issue that @idrarig is raising is that the user always writes to their own store server.

The issue is probably more clearly explained if you consider a Google Drive Upspin tree, where you could read and write your Google Drive content using the Upspin protocol (by way of a hypothetical `DriveServer`). To write a file to Google Drive in that instance:

- User puts blobs to the own StoreServer,
- User puts a DirEntry to DriveServer,
- DriveServer fetches the blobs from the user's StoreServer,
- DriveServer writes those blobs to Google Drive,
- DriveServer provides a DirEntry for that file in the tree, but its content is now provided by DriveServer's StoreServer implementation.

The awkwardness is in steps 2 and 3. The presence of a cacheserver may mean that the DriveServer can't fetch the blobs from the user's StoreServer yet.

So the fundamental question here is whether we should allow users to write to other StoreServers, and if so how that might work.

**Comment by edwardrf** [comment]
In upspin documentation, it described the 3 part of servers:

> Each of these three services is itself distributed, so for instance a user root may be hosted on or split across multiple DirServer instances, which coordinate to provide a consistent view of the user's data.

The wording there are vague enough that it does not prevent a easy service provider implementation, it largely hinge on how the DirServers "corrdinate".

If each DirServer don't have to provide a complete picture of the user's data, instead, each of them is responsible only for some of them. In this case, service provider can easily provide a DirServer for the user to add to his own set of DirServers, and all the writing can be done directly to service provider's implementation of storage server. (From security stand point, the user should share the data to the service provider's public key).

For example, for user alice:

/   -> RootDirServer -> DefaultStorageServer
/services/twitter/ -> TwitterDirServer -> TwitterStorageServer
/services/googleDrive/ -> GoogleDriveDirServer -> GoogleDriveStoageServer

This implementation also has some side effect:
- It provides a way to segregate user's data to different storage service, 
- It provides a way to write data to other people's storage

In practice, even though not implemented yet, the word "coordination" actually means "mirroring":
> 	// Dirs is a slice of DirServer endpoints where the user's root directory may be located,
	// which should be contacted in order when trying to access the user's tree. Multiple
	// entries represent mirrors, which are as yet unimplemented. TODO.

As for the cache server issue, I think it is actually an implementation detail, if cache server use a different upload priority strategy, uploads should not be blocked by a previous large file change.
